### PR TITLE
Add typescript guideline and typescript-specific eslint plugins and fix issues

### PR DESCRIPTION
**Pull request by silverwind**
1. Add some general guidelines how to write our typescript code
2. Add `@typescript-eslint/eslint-plugin`, general typescript rules
3. Add `eslint-plugin-deprecation` to detect deprecated code
4. Fix all new lint issues that came up

### CODE REVIEW

**Review by delvh — APPROVED** [review]

File: docs/content/development/hacking-on-gitea.en-us.md [hunk 1]
```diff
@@ -191,6 +191,22 @@ Before committing, make sure the linters pass:
...
+Prefer to use type aliases because they can represent any type and are generally more flexible to use than interfaces.
+
+### Use separate type imports
+
+We use `verbatimModuleSyntax` so type imports and non-type imports from the same file must be split into two `import type` statements. This enables the typescript compiler to completely eliminate the type import statements during compilation.
+
+### Use `@ts-expect-error` instead of `@ts-ignore`
+
+Both annotations should be avoided, but if you have to use them, use `@ts-expect-error` because it will not leave ineffective statements after the issue is fixed.
+
```
Comment (wxiaoguang) [inline docs/content/development/hacking-on-gitea.en-us.md]: I think these sections should be in frontend guideline doc (and that doc needs to be updated)

Comment (silverwind) [reply docs/content/development/hacking-on-gitea.en-us.md]: Ah, I forgot we had a separate document for it.

**Comment by silverwind** [comment]
I think we should try to run the TS rules on JS before the migration because that will allow the migration to be pure renames. Not sure if that is really possible already but I will try. If the diff is to big, it will be in another PR.

**Comment by silverwind** [comment]
`@typescript-eslint` now runs on JS files as well and has detected 7 new issues, one of them being a regression from https://github.com/go-gitea/gitea/pull/31186, which I also fixed.

**Review by jolheiser — APPROVED** [review]

**Review by jolheiser — APPROVED** [review]
Second time's the charm?

**Comment by silverwind** [comment]
Next and final step in the typescript journey: https://github.com/go-gitea/gitea/pull/31559

**Pull request merged**